features/topology-map/plugins/org.opennms.features.topology.plugins.topo.graphml/src/main/java/org/opennms/features/topology/plugins/topo/graphml/status/GraphMLPropagateVertexStatusProvider.java

```diff
@@ -0,0 +1,195 @@
+
+    public GraphMLPropagateVertexStatusProvider(final String namespace,
+                                                final GraphMLMetaTopologyProvider provider,
+                                                final BundleContext bundleContext) {
+        this.namespace = Objects.requireNonNull(namespace);
+        this.provider = Objects.requireNonNull(provider);
+        this.bundleContext = Objects.requireNonNull(bundleContext);
+    }
+
+    @Override
+    public Map<? extends VertexRef, ? extends Status> getStatusForVertices(final VertexProvider vertexProvider,
+                                                                           final Collection<VertexRef> vertices,
+                                                                           final Criteria[] criteria) {
+
+        final LoopDetectionCriteria loopDetectionCriteria = (LoopDetectionCriteria) Iterables.tryFind(Arrays.asList(criteria),
+                                                                                                      c -> c instanceof LoopDetectionCriteria)
+                                                                                             .or(LoopDetectionCriteria::new);
+
+        // Build map from namespace to opposite vertices
+        final Multimap<String, VertexRef> oppositeVertices = HashMultimap.create();
+        for (final VertexRef sourceVertex : vertices) {
+            // Filter out loops
+            if (loopDetectionCriteria.contains(sourceVertex)) {
+                LOG.error("Loop detected with: {}:{}", sourceVertex.getNamespace(), sourceVertex.getId());
+                continue;
+            }
+
+            for (VertexRef targetVertex : this.provider.getOppositeVertices(sourceVertex)) {
+                oppositeVertices.put(targetVertex.getNamespace(), targetVertex);
+            }
+        }
+
+        // Find and call status provider for each namespace and get result per opposite vertex
+        final Map<VertexRef, Status> targetStatuses = Maps.newHashMap();
+        try {
+            final Collection<ServiceReference<StatusProvider>> statusProviderReferences = this.bundleContext.getServiceReferences(StatusProvider.class, null);
+            for (final ServiceReference<StatusProvider> statusProviderReference : statusProviderReferences) {
+                try {
+                    final StatusProvider statusProvider = bundleContext.getService(statusProviderReference);
+
+                    for (final Map.Entry<String, Collection<VertexRef>> e : oppositeVertices.asMap().entrySet()) {
+                        if (statusProvider.contributesTo(e.getKey())) {
+                            targetStatuses.putAll(statusProvider.getStatusForVertices(
+                                                  this.provider.getGraphProviderBy(e.getKey()),
+                                                  e.getValue(),
+                                                  new Criteria[]{loopDetectionCriteria.with(vertices)}));
```

Comment:
At the moment this will work, but may cause side-effects in the future.
We should merge the original criteria with the (created) loopDetectionCriteria.